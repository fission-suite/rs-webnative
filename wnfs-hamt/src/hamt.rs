use super::{KeyValueChange, Node, HAMT_VERSION};
use crate::Hasher;
use anyhow::Result;
use async_trait::async_trait;
use libipld::{serde as ipld_serde, Ipld};
use semver::Version;
use serde::{
    de::{DeserializeOwned, Error as DeError},
    ser::Error as SerError,
    Deserialize, Deserializer, Serialize, Serializer,
};
use sha3::Sha3_256;
use std::{collections::BTreeMap, hash::Hash, rc::Rc, str::FromStr};
use wnfs_common::{AsyncSerialize, BlockStore, Link};

//--------------------------------------------------------------------------------------------------
// Type Definitions
//--------------------------------------------------------------------------------------------------

/// Hash Array Mapped Trie (HAMT) is an implementation of an associative array that combines the characteristics
/// of a hash table and an array mapped trie.
///
/// This type wraps the actual implementation which can be found in the [`Node`](crate::private::Node).
///
/// # Examples
///
/// ```
/// use wnfs_hamt::Hamt;
///
/// let hamt = Hamt::<String, usize>::new();
/// println!("HAMT: {:?}", hamt);
/// ```
#[derive(Debug, Clone)]
pub struct Hamt<K, V, H = Sha3_256>
where
    H: Hasher,
{
    pub root: Rc<Node<K, V, H>>,
    pub version: Version,
}

//--------------------------------------------------------------------------------------------------
// Implementations
//--------------------------------------------------------------------------------------------------

impl<K, V, H: Hasher> Hamt<K, V, H> {
    /// Creates a new empty HAMT.
    ///
    /// # Examples
    ///
    /// ```
    /// use wnfs_hamt::Hamt;
    ///
    /// let hamt = Hamt::<String, usize>::new();
    /// println!("HAMT: {:?}", hamt);
    /// ```
    pub fn new() -> Self {
        Self {
            root: Rc::new(Node::default()),
            version: HAMT_VERSION,
        }
    }

    /// Creates a new `HAMT` with the given root node.
    ///
    /// # Examples
    ///
    /// ```
    /// use std::rc::Rc;
    /// use wnfs_hamt::{Hamt, Node};
    ///
    /// let hamt = Hamt::<String, usize>::with_root(Rc::new(Node::default()));
    ///
    /// println!("HAMT: {:?}", hamt);
    /// ```
    pub fn with_root(root: Rc<Node<K, V, H>>) -> Self {
        Self {
            root,
            version: HAMT_VERSION,
        }
    }

    /// Gets the difference between two HAMTs at the key-value level.
    ///
    /// # Examples
    ///
    /// ```
    /// use std::rc::Rc;
    /// use wnfs_hamt::{Hamt, Node};
    /// use wnfs_common::MemoryBlockStore;
    ///
    /// #[async_std::main]
    /// async fn main() {
    ///     let store = &mut MemoryBlockStore::default();
    ///
    ///     let main_hamt = Hamt::<String, usize>::with_root({
    ///         let mut node = Rc::new(Node::default());
    ///         node.set("foo".into(), 400, store).await.unwrap();
    ///         node.set("bar".into(), 500, store).await.unwrap();
    ///         node
    ///     });
    ///
    ///     let other_hamt = Hamt::<String, usize>::with_root({
    ///         let mut node = Rc::new(Node::default());
    ///         node.set("foo".into(), 200, store).await.unwrap();
    ///         node.set("qux".into(), 600, store).await.unwrap();
    ///         node
    ///     });
    ///
    ///     let diff = main_hamt.diff(&other_hamt, store).await.unwrap();
    ///
    ///     println!("diff: {:#?}", diff);
    /// }
    pub async fn diff(
        &self,
        other: &Self,
        store: &mut impl BlockStore,
    ) -> Result<Vec<KeyValueChange<K, V>>>
    where
        K: DeserializeOwned + Clone + Eq + Hash + AsRef<[u8]>,
        V: DeserializeOwned + Clone + Eq,
        H: Clone + 'static,
    {
        super::diff(
            Link::from(Rc::clone(&self.root)),
            Link::from(Rc::clone(&other.root)),
            store,
        )
        .await
    }

    async fn to_ipld<B: BlockStore + ?Sized>(&self, store: &mut B) -> Result<Ipld>
    where
        K: Serialize,
        V: Serialize,
    {
        Ok(Ipld::Map(BTreeMap::from([
            ("root".into(), self.root.to_ipld(store).await?),
            ("version".into(), ipld_serde::to_ipld(&self.version)?),
            ("structure".into(), ipld_serde::to_ipld("hamt")?),
        ])))
    }
}

#[async_trait(?Send)]
impl<K, V, H: Hasher> AsyncSerialize for Hamt<K, V, H>
where
    K: Serialize,
    V: Serialize,
{
    async fn async_serialize<S, B>(&self, serializer: S, store: &mut B) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
        B: BlockStore + ?Sized,
    {
        self.to_ipld(store)
            .await
            .map_err(SerError::custom)?
            .serialize(serializer)
    }
}

impl<'de, K, V> Deserialize<'de> for Hamt<K, V>
where
    K: DeserializeOwned,
    V: DeserializeOwned,
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        Ipld::deserialize(deserializer).and_then(|ipld| ipld.try_into().map_err(DeError::custom))
    }
}

impl<K, V> TryFrom<Ipld> for Hamt<K, V>
where
    K: DeserializeOwned,
    V: DeserializeOwned,
{
    type Error = String;

    fn try_from(ipld: Ipld) -> Result<Self, Self::Error> {
        match ipld {
            Ipld::Map(mut map) => {
                let root = Rc::new(
                    Node::<K, V>::deserialize(map.remove("root").ok_or("Missing root")?)
                        .map_err(|e| e.to_string())?,
                );

                let version = match map.get("version").ok_or("Missing version")? {
                    Ipld::String(v) => Version::from_str(v).map_err(|e| e.to_string())?,
                    _ => return Err("`version` is not a string".into()),
                };

                Ok(Self { root, version })
            }
            other => Err(format!("Expected `Ipld::Map`, got {other:#?}")),
        }
    }
}

impl<K, V, H: Hasher> Default for Hamt<K, V, H> {
    fn default() -> Self {
        Self::new()
    }
}

impl<K, V, H> PartialEq for Hamt<K, V, H>
where
    K: PartialEq,
    V: PartialEq,
    H: Hasher,
{
    fn eq(&self, other: &Self) -> bool {
        self.root == other.root && self.version == other.version
    }
}

//--------------------------------------------------------------------------------------------------
// Tests
//--------------------------------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use wnfs_common::{dagcbor, MemoryBlockStore};

    #[async_std::test]
    async fn hamt_can_encode_decode_as_cbor() {
        let store = &mut MemoryBlockStore::default();
        let root = Rc::new(Node::default());
        let hamt: Hamt<String, i32> = Hamt::with_root(root);

        let encoded_hamt = dagcbor::async_encode(&hamt, store).await.unwrap();
        let decoded_hamt = dagcbor::decode::<Hamt<String, i32>>(encoded_hamt.as_ref()).unwrap();

        assert_eq!(hamt, decoded_hamt);
    }
}
