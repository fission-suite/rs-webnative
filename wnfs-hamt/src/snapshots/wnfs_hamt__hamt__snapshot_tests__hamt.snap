---
source: wnfs-hamt/src/hamt.rs
expression: hamt
---
{
  "value": {
    "root": [
      {
        "/": {
          "bytes": "//8"
        }
      },
      [
        {
          "/": "bafyr4ieu6bkebm4wnidg35ht634g6crheexkmjgj7ydy4s4vciqwknvj2y"
        },
        {
          "/": "bafyr4ibf3xts4e5lyutfcbl3z6mr2ub3yeq7ztuvlubmj56lqy44khuesq"
        },
        [
          [
            [
              62,
              0,
              0,
              0
            ],
            "62"
          ],
          [
            [
              87,
              0,
              0,
              0
            ],
            "87"
          ],
          [
            [
              24,
              0,
              0,
              0
            ],
            "24"
          ]
        ],
        {
          "/": "bafyr4igat27wtxad3it2conbavlglndlgfrmpcksmcg2n5m7fwsotwfme4"
        },
        {
          "/": "bafyr4ie3nvpfqmedgi4tbeypikcxrh7vydzo3pr246ykb46ato5mgo7i4e"
        },
        {
          "/": "bafyr4ibdl2emldhhi3xtxqbgrnalzola35tqhpfutbtvjrru5v7aqtjrxy"
        },
        [
          [
            [
              38,
              0,
              0,
              0
            ],
            "38"
          ],
          [
            [
              4,
              0,
              0,
              0
            ],
            "4"
          ],
          [
            [
              63,
              0,
              0,
              0
            ],
            "63"
          ]
        ],
        {
          "/": "bafyr4igwgoz5ljcl3rvhkueualixhp45ihn5bogkn7emmkluels45gvuie"
        },
        {
          "/": "bafyr4ic2kzon4twlsilhhlssgxgn7chv22lsnt3juc3weeye3p5fsiod7i"
        },
        {
          "/": "bafyr4ifenfmnyyztpmhikclvlzuwiidtzs5vcryjdtf6kqbapojcjkzze4"
        },
        [
          [
            [
              58,
              0,
              0,
              0
            ],
            "58"
          ]
        ],
        {
          "/": "bafyr4ibhrdbjd6fiwf3y7ck3qxb7wwc3vejhvmsaoe2omujfqcwt27stx4"
        },
        {
          "/": "bafyr4ibfaexqjhfn557cmuyvmksgqikfgf3asoj4cmypi6zvgyugnl3cy4"
        },
        {
          "/": "bafyr4ifl5uglwufirm5uexvzs57j5hzlxu3ybzt3kbdwgvr2naouyey5km"
        },
        {
          "/": "bafyr4ie2cyh4gcjdlvg3qjxghbleviq5yobbivv3qhx74tl2wlw5xmhweu"
        },
        {
          "/": "bafyr4ibkwabgy76xwvjky5pqmx42kryeeu3weftx6hxtsilfdmsdyfwgyq"
        }
      ]
    ],
    "structure": "hamt",
    "version": "0.1.0"
  },
  "cbor": "o2Ryb290gkL//5DYKlglAAFxHiCU8FRAs5ZqBm308/b4bwonIS6mJMn+B45LlRIhZTap1tgqWCUAAXEeICXd5y4Tq8UmUQV7z5kdUDvBIfzOlV0CxPfLhjnFHoSUg4KEGD4AAABiNjKChBhXAAAAYjg3goQYGAAAAGIyNNgqWCUAAXEeIMCev2ncA9onoTmhBVZltGsxYseJUmCNpvWfLaTp2Kwn2CpYJQABcR4gm21eWDCDMjkwkw9ChXif9cDy7b4657Cg88CbusM76OHYKlglAAFxHiAjXojFjOdG7zvAJotAvLlg32cDvLSYZ1TGNO1+CE0xvoOChBgmAAAAYjM4goQEAAAAYTSChBg/AAAAYjYz2CpYJQABcR4g1jOz1aRL3Gp1UJQC0XO/nUHb0LjKb8jGKXQi5c6atEHYKlglAAFxHiBaVlzeTsuSFnOuUjXM34j11pcmz2mgt2ITBNv6WSHD+tgqWCUAAXEeIKRpWNxjM3sOhQl1XmlkIHPMu1FHCRzL5UAge5IkqzkngYKEGDoAAABiNTjYKlglAAFxHiAniMKR+Kixd4+JW4XD+1hbqRJ6skBxNOZRJYCtPX5Tv9gqWCUAAXEeICUBLwScre9+JlMVYqRoIUUxdgk5PBMw9Hs1Nihmr2LH2CpYJQABcR4gq+0Mu1CoiztCXrmXfp6fK703gOZ7UEdjVjpoHUwTHVPYKlglAAFxHiCaFg/DCSNdTbgm5jhWSqIdw4IUVruB7/5NerLt27D2JdgqWCUAAXEeICqwAmx/17VSrHXwZfmlRwQlN2IWd/HvOSFlGyQ8FsbEZ3ZlcnNpb25lMC4xLjBpc3RydWN0dXJlZGhhbXQ="
}
